### ISSUE TITLE
Error and result clarification

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#192]
two problems:
a real kernel error in the comprehensive demo, and
the summary falsely says “CUDA NOT AVAILABLE” even though CUDA ran for other tests.

user established rule:
do not create any cpu fallback whatsoever.

Here’s how to fix both.

1) Kernel error (Comprehensive demo)

Root cause
In _flash_attn_fwd, you write the attention mask tile with a block pointer and pass both boundary_check= and mask= to tl.store. Triton doesn’t allow mask= with block pointers. That’s what triggers the failure you see around:

tl.store(
    output_attn_mask_tile_ptr,
    mask.to(tl.int8),
    boundary_check=(0, 1),
    mask=boundary_mask,   # ❌ not allowed with block_ptr
)

Fix
Pre-apply the boundary mask and store without mask=:

# Before:
boundary_mask = (q_tile_indices[:, None] < seq_len) & (kv_indices[None, :] < seq_len)
tl.store(
    output_attn_mask_tile_ptr,
    mask.to(tl.int8),
    boundary_check=(0, 1),
    mask=boundary_mask,
)

# After:
boundary_mask = (q_tile_indices[:, None] < seq_len) & (kv_indices[None, :] < seq_len)
safe_mask_tile = (mask & boundary_mask).to(tl.int8)
tl.store(
    output_attn_mask_tile_ptr,
    safe_mask_tile,
    boundary_check=(0, 1),
)

(Do the same pattern anywhere else you try to combine boundary_check and mask on a block pointer.)

This will make test_comprehensive_attention_demonstration actually execute the kernel instead of bailing out.

2) Misleading “CUDA NOT AVAILABLE” summary

Root cause
Your skip branches always message “CUDA unavailable” and your final summary prints a hardcoded header:

print(f"⏭️  CUDA TESTS SKIPPED (...) - CUDA NOT AVAILABLE:")

So a kernel failure gets misreported as “CUDA not available”.

Fix A (make the skip reason accurate)
When a kernel call fails (but CUDA exists), change the skip reason/messages to “kernel execution failed”, not “CUDA unavailable”.

In these tests:

test_comprehensive_attention_demonstration

test_kernel_attention_patterns_teacher_forcing

test_kernel_attention_patterns_cocktail_party

replace the else branch:

print("⚠️  CUDA kernel execution required but not available!")
...
self.skipTest("CUDA kernel execution required for ... but CUDA unavailable")

with:

print("⚠️  CUDA kernel execution failed; cannot validate without a working kernel.")
self.skipTest("CUDA kernel execution failed for attention pattern validation")

Fix B (make the footer header truthful)
In run_tests() where you print the categorization, choose the header based on the actual skip reasons:

if result.cuda_skipped:
    reasons = [reason for _, reason in result.cuda_skipped]
    unavailable = any("unavailable" in reason.lower() or "not available" in reason.lower() for reason in reasons)
    header = "CUDA NOT AVAILABLE" if unavailable else "KERNEL EXECUTION FAILED"
    print(f"⏭️  CUDA TESTS SKIPPED ({len(result.cuda_skipped)}) - {header}:")
    for test_name, reason in result.cuda_skipped:
        print(f"   • {test_name}")
        print(f"     Reason: {reason}")

This way:

if you really skipped due to no CUDA, it says so,

if you skipped because the kernel errored, the summary says “KERNEL EXECUTION FAILED”.

============================================================
ATTENTION TOKEN BEHAVIOR TESTS
============================================================
Testing attention patterns for teacher forcing and cocktail party tasks
Note: These tests validate the attention logic without running Triton kernels

test_attention_mask_creation (__main__.AttentionBehaviorTests.test_attention_mask_creation)
Test creation of attention masks for different patterns. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Attention Mask Creation ===
Sequence length: 16
Is prefix: tensor([ True,  True,  True,  True, False, False, False, False, False, False,
        False, False, False, False, False, False])
In span: tensor([False, False, False, False, False, False, False, False,  True,  True,
         True,  True,  True,  True,  True, False])
Span ID: tensor([ 0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  2,  2,  2, -1])
Prefix positions: [0, 1, 2, 3]
Context positions: [4, 5, 6, 7]
Span 1 positions: [8, 9, 10, 11]
Span 2 positions: [12, 13, 14]
MASKQ positions: [15]
ok
test_attention_pattern_logic_validation (__main__.AttentionBehaviorTests.test_attention_pattern_logic_validation)
Test that the attention pattern logic itself is correctly implemented. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Attention Pattern Logic Validation ===
Test setup:
  is_prefix: tensor([ True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False])
  in_span: tensor([False, False, False, False, False,  True,  True,  True,  True,  True,
         True, False, False, False, False, False, False, False, False, False])
  span_id: tensor([ 0,  0,  0,  0,  0,  1,  1,  1,  2,  2,  2, -1,  0,  0,  0,  0,  0,  0,
         0,  0])

Validating attention pattern logic:
1. Prefix positions: [0, 1, 2]
   ✓ Prefix tokens should attend bidirectionally to each other
2. Context positions: [3, 4]
   ✓ Context tokens should attend causally + see prefix
3. Span 1 positions: [5, 6, 7]
   Span 2 positions: [8, 9, 10]
   ✓ Spans should be bidirectional within span, see context, not see each other
4. MASKQ positions: [11]
   ✓ MASKQ should see all spans and prefix
ok
test_comprehensive_attention_demonstration (__main__.AttentionBehaviorTests.test_comprehensive_attention_demonstration)
Comprehensive test that demonstrates all attention behaviors clearly. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Comprehensive Attention Pattern Demonstration ===
This test demonstrates the complete cocktail party attention behaviors
as described in the issue, with clear visualization of patterns.

🎭 Test Sequence Design:
  • Positions 0-2: Prefix tokens (including CLS at 2)
  • Positions 3-4: Context tokens
  • Positions 5-8: Span 1 ([SPAN] content content [ES])
  • Positions 9-11: Span 2 ([SPAN] content [ES])
  • Position 12: MASKQ token
  • Positions 13+: PAD tokens

🔧 Tensor Setup:
  • Device: cuda
  • Input shapes: q=torch.Size([1, 2, 20, 32]), k=torch.Size([1, 2, 20, 32]), v=torch.Size([1, 2, 20, 32])
  • is_prefix: tensor([ True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False])
  • in_span: tensor([False, False, False, False, False,  True,  True,  True,  True,  True,
         True,  True, False, False, False, False, False, False, False, False])
  • span_id: tensor([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  2,  2,  2, -1,  0,  0,  0,  0,  0,
         0,  0])
❌ Comprehensive Demo: CUDA kernel failed (at 272:16:
                order=(1, 0),
            )

            # Write the mask to the output tensor
            if Q_BLOCK_DIVISIBLE and K_BLOCK_DIVISIBLE:
                tl.store(
                    output_attn_mask_tile_ptr,
                    mask.to(tl.int8),  # Convert bool to int8 for storage
                )
            else:
                boundary_mask = (q_tile_indices[:, None] < seq_len) & (kv_indices[None, :] < seq_len)
                tl.store(
                ^)
   Test requires CUDA execution - skipping to avoid misleading results

⚠️  CUDA kernel execution required but not available!
   Cannot demonstrate cocktail party behaviors without actual kernel execution.
   Skipping test as it requires CUDA kernel execution.
skipped 'CUDA kernel execution required for attention pattern validation but CUDA unavailable'
test_data_builder_cocktail_party_format (__main__.AttentionBehaviorTests.test_data_builder_cocktail_party_format)
Test data builder creates proper cocktail party format. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Data Builder Cocktail Party Format ===
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 5 samples from the dataset.
Loading dataset: allenai/c4/en
Attempting Method 1: Load C4 'en' (streaming)...
Resolving data files: 100%|████████████████| 1024/1024 [00:03<00:00, 309.63it/s]
Resolving data files: 100%|█████████████| 1024/1024 [00:00<00:00, 164425.84it/s]
C4 'en' (streaming) load_dataset call succeeded. Processing samples...
Reached max_samples (5) for C4 'en' streaming.
Finished processing C4 'en' streaming. Total valid samples extracted: 5
Successfully processed 5 samples from C4 'en' stream.
Successfully loaded 5 samples using C4 'en'.
Final dataset loaded with 5 samples.
Returning dataset splits: train=4, val=1, test=1
Tokenizing train split...
Text length for train: 3519 characters
Tokenized to 3519 byte tokens
Tokenizing validation split...
Text length for validation: 748 characters
Tokenized to 748 byte tokens
Tokenizing test split...
Text length for test: 1014 characters
Tokenized to 1016 byte tokens
train dataset: 3455 samples
Limited validation to 129 tokens for faster evaluation (target: 129)
validation dataset: 65 samples
Limited test to 129 tokens for faster evaluation (target: 129)
test dataset: 65 samples
✅ Cocktail party dataloader created successfully
   Batch inputs shape: torch.Size([8, 64])
   Correct indices: tensor([1, 1, 1, 1, 1, 0, 0, 0])
   Contains CLS: False
   Contains SPAN: True
   Contains ES: True
   Contains MASKQ: True
   Sample text (first 32 tokens): ou let CCC do the clone, the[MASK]sul
   ⚠️  CLS token not found (may be using test data without task prefixes)
   ✅ Span tokens found - cocktail party structure present
✅ Data builder cocktail party format validated successfully!
ok
test_kernel_attention_patterns_cocktail_party (__main__.AttentionBehaviorTests.test_kernel_attention_patterns_cocktail_party)
Test that the actual kernel's attention patterns match expected behavior for cocktail party. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Kernel Attention Patterns - Cocktail Party ===
Test sequence shape: torch.Size([2, 32])
Prefix mask: tensor([ True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
In span mask: tensor([False, False, False, False, False, False, False,  True,  True,  True,
         True,  True,  True,  True,  True,  True, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
Span IDs: tensor([ 0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2, -1,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0])
Input tensor shapes: q=torch.Size([2, 4, 32, 16]), k=torch.Size([2, 4, 32, 16]), v=torch.Size([2, 4, 32, 16])
Tensors created on device: cuda:0
✅ Cocktail Party: Successfully ran CUDA kernel
🎯 Successfully obtained attention mask from CUDA kernel: torch.Size([2, 4, 32, 32])

🎭 Cocktail Party Attention Pattern Analysis:
  🔍 Testing prefix bidirectional attention...
  ✅ Prefix bidirectional attention: PASSED
  🔍 Testing span isolation...
  ✅ Span isolation: PASSED
  🔍 Testing MASKQ visibility...
  ✅ MASKQ visibility: PASSED
  🔍 Testing context cocktail party behavior...
  ✅ Context cocktail party behavior: PASSED

📊 Detailed Attention Pattern Analysis:
  📍 Token Analysis for Batch 0:
    • Prefix positions: [0, 1, 2]
    • Context positions: [3, 4, 5, 6, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]
    • Span 1 positions: [7, 8, 9, 10, 11]
    • Span 2 positions: [12, 13, 14, 15]
    • MASKQ positions: [16]

  🔍 Sample Attention Patterns (Head 0):
    • Prefix token 0 attends to: [0, 1, 2]
    • Context token 3 attends to: [0, 1, 2, 3]
    • Span 1 token 7 attends to: [3, 4, 5, 6, 7, 8, 9, 10, 11, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]
    • Span 2 token 12 attends to: [3, 4, 5, 6, 12, 13, 14, 15, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]
    • MASKQ token 16 attends to: [0, 1, 2, 7, 8, 9, 10, 11, 12, 13, 14, 15]
  📈 This demonstrates the cocktail party attention isolation and visibility patterns!

✅ All cocktail party attention patterns validated from real kernel!
🔧 API signature correctly accepts cocktail party metadata parameters
ok
test_kernel_attention_patterns_teacher_forcing (__main__.AttentionBehaviorTests.test_kernel_attention_patterns_teacher_forcing)
Test that the actual kernel's attention patterns match expected behavior for teacher forcing. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Kernel Attention Patterns - Teacher Forcing ===
Test sequence shape: torch.Size([2, 32])
CLS position: 3
Prefix mask: tensor([ True,  True,  True,  True, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
Input tensor shapes: q=torch.Size([2, 4, 32, 16]), k=torch.Size([2, 4, 32, 16]), v=torch.Size([2, 4, 32, 16])
Tensors created on device: cuda:0
✅ Teacher Forcing: Successfully ran CUDA kernel
🎯 Successfully obtained attention mask from CUDA kernel: torch.Size([2, 4, 32, 32])
Validating kernel-generated attention mask for teacher forcing...
✓ Kernel attention mask validation completed
  🔍 Testing prefix bidirectional attention...
  ✅ Prefix bidirectional attention: PASSED
  🔍 Testing context causal attention...
  ✅ Context causal attention: PASSED
✅ All teacher forcing attention patterns validated from real kernel!
🔧 API signature correctly accepts teacher forcing parameters
ok
test_kernel_mask_output_integration (__main__.AttentionBehaviorTests.test_kernel_mask_output_integration)
Test that demonstrates the complete integration of kernel mask output functionality. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Kernel Mask Output Integration ===
This test validates the complete solution to issue #177:
1. ✓ Modified flash_attention to accept return_attention_mask parameter
2. ✓ Updated kernel signature to support attention mask output
3. ✓ Added logic to write computed masks to output tensor
4. ✓ Made functionality toggleable via optional argument
5. ✓ Maintained existing kernel behavior when not requested

Test 1: Default behavior (no mask output)
✓ Expected CUDA requirement confirmed

Test 2: Mask output behavior
✓ Expected CUDA requirement confirmed for mask output

Test 3: Parameter validation
✓ All required parameters can be passed to flash_attention
✓ Cocktail party metadata integrated into kernel API
✓ return_attention_mask parameter controls output format

Test 4: Kernel modifications summary
✓ Added OUTPUT_ATTN_MASK tensor parameter to Triton kernel
✓ Added mask writing logic in attention computation loop
✓ Added RETURN_ATTENTION_MASK compile-time constant
✓ Kernel writes computed mask when requested
✓ No performance impact when mask output not requested

✓ Integration test completed successfully!
  Note: Actual CUDA execution would demonstrate mask correctness
  This test validates the complete API and integration changes
ok
test_special_token_behaviors (__main__.AttentionBehaviorTests.test_special_token_behaviors)
Test special token handling behaviors. ... 
🖥️  GPU: NVIDIA H100 80GB HBM3 (Compute Capability: 9.0)
🔧 Device: cuda
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Special Token Behaviors ===
1. Testing PAD token handling...
   Sequence: tensor([10, 11,  1, 12,  0,  0,  0,  0])
   Attention mask: tensor([ True,  True,  True,  True, False, False, False, False])
   Valid length: 4
2. Testing CLS token detection...
   CLS token found at position: 2
3. Testing MASK token in context...
   MASK token found at position: 3
   MASK token should behave like regular context token (causal)
✓ All special token behaviors validated successfully!
ok

----------------------------------------------------------------------
Ran 8 tests in 9.885s

OK (skipped=1)

============================================================
TEST RESULTS CATEGORIZATION
============================================================
⏭️  CUDA TESTS SKIPPED (1) - CUDA NOT AVAILABLE:
   • AttentionBehaviorTests.test_comprehensive_attention_demonstration
     Reason: CUDA kernel execution required for attention pattern validation but CUDA unavailable

============================================================
⏭️  1 CUDA KERNEL TESTS SKIPPED!
CUDA tests were skipped because CUDA is not available.
To properly test attention kernels, run on a CUDA-enabled device.
============================================================
